### PR TITLE
Specifying environment variables in RunSettings file

### PR DESCRIPTION
**Pull request by vagisha-nidhi**
## Description
Option to specify environment variables in run settings file. 
RFC: [https://github.com/microsoft/vstest-docs/pull/202](https://github.com/microsoft/vstest-docs/pull/202)

For example:
```
<?xml version="1.0" encoding="utf-8"?>
<!-- File name extension must be .runsettings -->
<RunSettings>
  <RunConfiguration>
    <EnvironmentVariables>
      <!-- List of environment variables we want to set-->
      <DOTNET_ROOT>C:\ProgramFiles\dotnet</DOTNET_ROOT>
      <SDK_PATH>C:\Codebase\Sdk</SDK_PATH>
    </EnvironmentVariables>
  </RunConfiguration>
</RunSettings>
```
These are set when spawning the test host process.

## Related issue
[https://github.com/microsoft/vstest/issues/2100](https://github.com/microsoft/vstest/issues/2100)

### CODE REVIEW

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyOperationManager.cs [hunk 1]
```diff
@@ -25,7 +25,8 @@ namespace Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Client
     using CrossPlatEngineResources = Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Resources.Resources;
     using CommunicationUtilitiesResources = Microsoft.VisualStudio.TestPlatform.CommunicationUtilities.Resources.Resources;
     using CoreUtilitiesConstants = Microsoft.VisualStudio.TestPlatform.CoreUtilities.Constants;
-   
+    using System.Collections;
```
Comment (singhsarab) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyOperationManager.cs]: Remove

File: src/vstest.console/Processors/RunSettingsArgumentProcessor.cs [hunk 2]
```diff
@@ -137,6 +138,12 @@ public void Initialize(string argument)
                 this.runSettingsManager.AddDefaultRunSettings();
 
                 this.commandLineOptions.SettingsFile = argument;
+
+                if (this.runSettingsManager.QueryRunSettingsNode("RunConfiguration.EnvironmentVariables") != null)
```
Comment (singhsarab) [inline src/vstest.console/Processors/RunSettingsArgumentProcessor.cs]: >"RunConfiguration.EnvironmentVariables" [](start = 65, length = 39)

Constants ?

File: src/vstest.console/Processors/RunSettingsArgumentProcessor.cs [hunk 3]
```diff
@@ -137,6 +138,12 @@ public void Initialize(string argument)
...
+                {
+                    this.commandLineOptions.InIsolation = true;
+                    this.runSettingsManager.UpdateRunSettingsNode(InIsolationArgumentExecutor.RunSettingsPath, "true");
```
Comment (singhsarab) [inline src/vstest.console/Processors/RunSettingsArgumentProcessor.cs]: Just curious, why is this required if the CLI option is already set.

File: test/Microsoft.TestPlatform.AcceptanceTests/RunsettingsTests.cs [hunk 4]
```diff
@@ -447,6 +447,37 @@ public void LegacySettingsAssemblyResolution(RunnerInfo runnerInfo)
...
+        public void EnvironmentVariablesSettingsShouldSetEnvironmentVariables(RunnerInfo runnerInfo)
+        {
+            AcceptanceTestBase.SetTestEnvironment(this.testEnvironment, runnerInfo);
+
+            var testAssemblyPath = this.GetAssetFullPath("EnvironmentVariablesTestProject.dll");
+
+            var runsettingsXml = @"<RunSettings>
+                                    <RunConfiguration>
+                                      <EnvironmentVariables>
+                                        <DOTNET_ROOT>C:\ProgramFiles\dotnet</DOTNET_ROOT>
```
Comment (singhsarab) [inline test/Microsoft.TestPlatform.AcceptanceTests/RunsettingsTests.cs]: Diff env please.

File: test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/DataCollection/ProxyDataCollectionManagerTests.cs [hunk 5]
```diff
@@ -144,6 +144,26 @@ public void InitializeShouldPassDiagArgumentsIfDiagIsEnabled()
             }
         }
 
+        [TestMethod]
+        public void InitilizeShouldLaunchDataCollectorWithEnvironmentVariables()
```
Comment (singhsarab) [inline test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/DataCollection/ProxyDataCollectionManagerTests.cs]: initialize*

**Comment by singhsarab** [comment]
>         private class TestableRunSettingsArgumentExecutor : RunSettingsArgumentExecutor

FileHelper please

---
Refers to: test/vstest.console.UnitTests/Processors/RunSettingsArgumentProcessorTests.cs:354 in 9ac75ee. [](commit_id = 9ac75ee971ffd4d170e3f57490929bef96b33774, deletion_comment = False)

File: test/Microsoft.TestPlatform.Utilities.UnitTests/InferRunSettingsHelperTests.cs [hunk 6]
```diff
@@ -558,6 +558,51 @@ public void TryGetLegacySettingsForRunSettingsWithValidLegacySettingsShouldRetur
...
+            var envVars = InferRunSettingsHelper.GetEnvironmentVariables(runSettingsXml);
+            Assert.AreEqual(0, envVars.Count);
+        }
+
+        [TestMethod]
+        public void GetEnvironmentVariablesWithInvalidValuesInRunSettingsShouldReturnEmptyDictionary()
+        {
+            string runSettingsXml = @"<RunSettings>
+                                       <RunConfiguration>
+                                         <Foo>
```
Comment (singhsarab) [inline test/Microsoft.TestPlatform.Utilities.UnitTests/InferRunSettingsHelperTests.cs]: <EnvironmentVariables><Foo></EnvironmentVariables>

File: src/Microsoft.TestPlatform.Utilities/InferRunSettingsHelper.cs [hunk 7]
```diff
@@ -377,6 +379,40 @@ private static List<string> GetNodeAttributes(XPathNavigator node)
...
+                    if (node == null)
+                    {
+                        return environmentVariables;
+                    }
+
+                    var childNodes = node.SelectChildren(XPathNodeType.Element);
+
+                    while (childNodes.MoveNext())
+                    {
+                        environmentVariables.Add(childNodes.Current.Name, childNodes.Current.Value);
```
Comment (singhsarab) [inline src/Microsoft.TestPlatform.Utilities/InferRunSettingsHelper.cs]: Checks

File: src/Microsoft.TestPlatform.Utilities/InferRunSettingsHelper.cs [hunk 8]
```diff
@@ -377,6 +379,40 @@ private static List<string> GetNodeAttributes(XPathNavigator node)
             return null;
         }
 
+        public static Dictionary<string, string> GetEnvironmentVariables(string runSettings)
```
Comment (singhsarab) [inline src/Microsoft.TestPlatform.Utilities/InferRunSettingsHelper.cs]: Comments

File: src/Microsoft.TestPlatform.Utilities/InferRunSettingsHelper.cs [hunk 9]
```diff
@@ -377,6 +379,40 @@ private static List<string> GetNodeAttributes(XPathNavigator node)
...
+        {
+            var environmentVariables = new Dictionary<string, string>();
```
Comment (singhsarab) [inline src/Microsoft.TestPlatform.Utilities/InferRunSettingsHelper.cs]: move below the node null check

File: src/vstest.console/Processors/RunSettingsArgumentProcessor.cs (hunk 3 above)
Comment (vagisha-nidhi) [reply src/vstest.console/Processors/RunSettingsArgumentProcessor.cs]: [https://github.com/microsoft/vstest/blob/master/src/Microsoft.TestPlatform.CrossPlatEngine/TestEngine.cs#L244](https://github.com/microsoft/vstest/blob/master/src/Microsoft.TestPlatform.CrossPlatEngine/TestEngine.cs#L244) This checks inisolation node from runsettings.

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyOperationManager.cs (hunk 1 above)
Comment (vagisha-nidhi) [reply src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyOperationManager.cs]: done

File: src/vstest.console/Processors/RunSettingsArgumentProcessor.cs (hunk 2 above)
Comment (vagisha-nidhi) [reply src/vstest.console/Processors/RunSettingsArgumentProcessor.cs]: Not available

File: src/Microsoft.TestPlatform.Utilities/InferRunSettingsHelper.cs (hunk 8 above)
Comment (vagisha-nidhi) [reply src/Microsoft.TestPlatform.Utilities/InferRunSettingsHelper.cs]: Added

File: src/Microsoft.TestPlatform.Utilities/InferRunSettingsHelper.cs (hunk 7 above)
Comment (vagisha-nidhi) [reply src/Microsoft.TestPlatform.Utilities/InferRunSettingsHelper.cs]: This comes out to be always empty and never null.
The following is invalid  and document fails to load.
```
<RunConfiguration>
 <EnvironmentVariables>
   <Foo>
 </EnvironmentVariables>
</RunConfiguration>
```

File: test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/DataCollection/ProxyDataCollectionManagerTests.cs (hunk 5 above)
Comment (vagisha-nidhi) [reply test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/DataCollection/ProxyDataCollectionManagerTests.cs]: Done

File: test/Microsoft.TestPlatform.Utilities.UnitTests/InferRunSettingsHelperTests.cs [hunk 10]
```diff
@@ -558,6 +558,66 @@ public void TryGetLegacySettingsForRunSettingsWithValidLegacySettingsShouldRetur
...
         }
 
+        [TestMethod]
+        public void GetEnvironmentVariablesWithValidValuesInRunSettingsShouldReturnValidDictionary()
+        {
+            string runSettingsXml = @"<RunSettings>
+                                       <RunConfiguration>
+                                          <EnvironmentVariables>
+                                             <RANDOM_PATH>C:\temp</RANDOM_PATH>
+                                             <RANDOM_PATH2>C:\temp2</RANDOM_PATH2>
```
Comment (singhsarab) [inline test/Microsoft.TestPlatform.Utilities.UnitTests/InferRunSettingsHelperTests.cs]: Check for the case when user tried to add the same key twice

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/ProxyDataCollectionManager.cs [hunk 11]
```diff
@@ -226,7 +226,7 @@ public void Initialize()
             this.dataCollectionPort = this.dataCollectionRequestSender.InitializeCommunication();
 
             // Warn the user that execution will wait for debugger attach.
-            this.dataCollectionProcessId = this.dataCollectionLauncher.LaunchDataCollector(null, this.GetCommandLineArguments(this.dataCollectionPort));
+            this.dataCollectionProcessId = this.dataCollectionLauncher.LaunchDataCollector(InferRunSettingsHelper.GetEnvironmentVariables(SettingsXml), this.GetCommandLineArguments(this.dataCollectionPort));
             EqtTrace.Info("ProxyDataCollectionManager.Initialize: Launched datacollector processId: {0} port: {1}", this.dataCollectionProcessId, this.dataCollectionPort);
```
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/ProxyDataCollectionManager.cs]: >acollector processId: {0} port: {1}", this.dataCollectionProcessId, this.dataColl [](start = 78, length = 81)

Log env vars used to launch process

File: src/Microsoft.TestPlatform.Utilities/InferRunSettingsHelper.cs [hunk 12]
```diff
@@ -377,6 +379,50 @@ private static List<string> GetNodeAttributes(XPathNavigator node)
...
+                    {
+                        return null;
+                    }
+
+                    environmentVariables = new Dictionary<string, string>();
+                    var childNodes = node.SelectChildren(XPathNodeType.Element);
+
+                    while (childNodes.MoveNext())
+                    {
+                        if (!environmentVariables.ContainsKey(childNodes.Current.Name))
```
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.Utilities/InferRunSettingsHelper.cs]: >ContainsKey [](start = 50, length = 11)

You can use TryAdd, instead of checking and then adding.

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/ProxyDataCollectionManager.cs [hunk 13]
```diff
@@ -226,7 +226,7 @@ public void Initialize()
             this.dataCollectionPort = this.dataCollectionRequestSender.InitializeCommunication();
 
             // Warn the user that execution will wait for debugger attach.
-            this.dataCollectionProcessId = this.dataCollectionLauncher.LaunchDataCollector(null, this.GetCommandLineArguments(this.dataCollectionPort));
+            this.dataCollectionProcessId = this.dataCollectionLauncher.LaunchDataCollector(InferRunSettingsHelper.GetEnvironmentVariables(SettingsXml), this.GetCommandLineArguments(this.dataCollectionPort));
```
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/ProxyDataCollectionManager.cs]: IMO this is not needed, & can potentially cause issues where certain Env variables, like Profiler variables if specified in runsettings would also get applied on top of datacollector, which can potentially cause data collector to be started under profiling mode, I'm not sure how will those scenarios be handled.

We can remove this for now, & add it only when needed.

**Review by mayankbansal018 — APPROVED** [review]
:shipit:

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/ProxyDataCollectionManager.cs (hunk 13 above)
Comment (vagisha-nidhi) [reply src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/ProxyDataCollectionManager.cs]: done

**Comment by ViktorHofer** [comment]
Very happy to see this implemented. Please cc me the next time in the implementation PR as well. Were you able to address my comment (https://github.com/microsoft/vstest-docs/pull/202#discussion_r312434371) as well?

**Comment by singhsarab** [comment]
@ViktorHofer There is no scenario where we would now be running in vstest.console, wherenever there are environment variables specified, it runs in isolation which means testhost will be spawned.
